### PR TITLE
Correct handling of whitespace in base64 URLs

### PR DESCRIPTION
**Pull request by pshaughn**
Whitespace handling in base64 url parsing now treats tabs, form-feeds, and linefeeds the same way it treats regular spaces

---

- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes fix (part of) #2909

- [X] There are tests for these changes OR
- [ ] These changes do not require tests because ___

### CODE REVIEW

File: components/net/data_loader.rs [hunk 1]
```diff
@@ -44,11 +44,13 @@ pub fn decode(url: &ServoUrl) -> Result<DecodeData, DecodeError> {
...
-        // FIXME(#2909): It’s unclear what to do with non-alphabet characters,
-        // but Acid 3 apparently depends on spaces being ignored.
+        // FIXME(#2909):
+        // infra.spec.whatwg.org/#forgiving-base64-decode does a pretty good
+        // job of telling us what to do here. Most WPT tests are working;
+        // the base64 string "ab=" in particular is not.
         bytes = bytes
             .into_iter()
-            .filter(|&b| b != b' ')
+            .filter(|&b| b != b' ' && b != b'\x0C' && b != b'\n' && b != b'\r' && b != b'\t')
```
Comment (jdm) [inline components/net/data_loader.rs]: While it's logical to extend the existing code here, we should honour SimonSapin's request in https://github.com/servo/servo/issues/2909#issuecomment-558328602 to make use of the [data-url crate](https://docs.rs/data-url/0.1.0/data_url/) and replace all of this decoding code if possible.

We should be able to use `DataUrl::process(url.as_str())` and return the value of decode_to_vec() along with the url's mime_type() value. We'll need to add data-url as a new dependency in components/net/Cargo.toml as well.

**Review by jdm — CHANGES_REQUESTED** [review]
Good find on the data url decoding problems!

File: components/net/data_loader.rs (hunk 1 above)
Comment (pshaughn) [reply components/net/data_loader.rs]: Will do! I've never added a Cargo dependency before, but I'll have to sooner or later and I may as well start with this one!

**Comment by pshaughn** [comment]
Now using data-url's version of forgiving base64 decoding, but still doing the rest of the MIME parse within data_loader.rs, as suggested in https://github.com/servo/servo/issues/2909#issuecomment-558537093. This is a full fix for #2909 as far as WPT tests are concerned, but integrating with the full data-url class eventually would be nice.

**Comment by SimonSapin** [comment]
Could you squash your commits into one? Assuming https://github.com/servo/servo is your `origin` remote, that would be something like `git fetch origin && git rebase -i origin/master`. Or I can do it for you prefer.

**Comment by pshaughn** [comment]
@SimonSapin : Is there a way to do that within this pull request, or should I make a new pull request with the merged commit?

**Comment by pshaughn** [comment]
#24870 is the commit-squashed version of this, sorry for the mess

**Pull request closed without merging**